modules/frontend/combiner/metrics_query_range.go

```diff
@@ -43,6 +44,11 @@ func NewQueryRange(req *tempopb.QueryRangeRequest, trackDiffs bool) (Combiner, e
 			if resp == nil {
 				resp = &tempopb.QueryRangeResponse{}
 			}
+			if setMaxSeries && len(resp.Series) > maxSeries {
+				resp.Series = resp.Series[:maxSeries]
```

Comment:
IIUC, we are still collecting all the series and the dropping the extra data before we return them, and not exiting early here? right?

it would be great if exited early when we hit this limit.

It would be very useful in the cases where is q metrics query is returning high cardinality results, for example: `{} | rate by (span:id)`

Just the work of pulling the series from the blocks will be resource intensive, and can OOM all generators.

Reply:
i raised the question here https://github.com/grafana/tempo/pull/4525#issuecomment-2605057481
the problem is the results come in indeterministically - so if we exit as soon as we hit the max series, we could have a response where each series just has one data point which isn't very uesful. but yes it would save us for out performing.

Reply:
cool, I am okay with current impl, and we can leve a note or todo that this can improved by making the results deterministic and exiting early.

sidenote: I think Joe's ordered results work might help here.